notifier.cpp

```diff
@@ -44,15 +55,38 @@ std::string get_subject(const std::string& data) {
     return subject;
 }
 
+bool is_file_empty(std::ifstream& file) {
+    return file.tellg() == std::ifstream::traits_type::eof();
+}
+
+void fill_backup_msgs() {
+    std::ifstream backup(backup_file);
+    if (backup.is_open()) {
+        if (is_file_empty(backup))
+            return;
+
+        for (auto [i, line_data, line_subject] = std::tuple<uint64_t, std::string, std::string>{0, "", ""};
+             std::getline(backup, line_subject), std::getline(backup, line_data); i++) {
+            msgs_queue.insert({ i, {line_subject, line_data} });
+        }
+        backup.close();
+        std::remove(backup_file.c_str());
+    }
+}
+
 static void _publish_ack_cb(const char* guid, const char* error, void* closure) {
-    std::unique_ptr<message> msg(static_cast<message*>(closure));
     // TODO: delete object from waiting list, so we can check if some object didn't published for a long time
     //std::cout << "#Ack#, " << guid << std::endl;
     // myPubMsgInfo* pubMsg = (myPubMsgInfo*)closure;
     // printf("Ack handler for message ID=%s Data=%.*s GUID=%s - ", pubMsg->ID, pubMsg->size, pubMsg->payload, guid);
+
     if (error != NULL) {
         std::cout << "Error: " << error << std::endl;
+        bad_msgs_queue.push_back(*(static_cast<uint64_t*>(closure)));
         done = true;    // TODO: locking
+    } else {
+        auto index = *(static_cast<uint64_t*>(closure));
```

Comment:
It is not safe to pass pointer here.
Better way, pass here value of index `auto index = static_cast<uint64_t>(closure);`